File: python/19-Remove Nth Node From End of List.py
from linklist_helper import *

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        "Time: O(n), Space: O(1)"
        dummy = ListNode()
        dummy.next = head
        l, r = dummy, dummy
        for _ in range(n+1):
            r = r.next
        while r:
            l = l.next
            r = r.next
        l.next = l.next.next
        return dummy.next

head = gen_linklist([1,2,3,4,5])
n = 2
result = Solution().removeNthFromEnd(head, n)
display_linklist(result) # [1,2,3,5]

head = gen_linklist([1])
n = 1
result = Solution().removeNthFromEnd(head, n)
display_linklist(result) # []

head = gen_linklist([1,2])
n = 1
result = Solution().removeNthFromEnd(head, n)
display_linklist(result) # [1]

File: cpp/19. Remove Nth Node From End of List.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode dummy;
        dummy.next = head;
        ListNode *l = &dummy, *r = &dummy;       
        for(int i = 0; i < n+1; i++) r = r->next;
        while(r) {
            l = l->next;
            r = r->next;
        }
        l->next = l->next->next;
        return dummy.next;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    
}